// @lodestar/utils/lib/assert.js
export const assert = {
    /**
     * Assert condition is true, otherwise throw AssertionError
     */
    true(condition, message) {
        if (!condition) {
            throw new AssertionError(message || "Expect condition to be true");
        }
    },
    /**
     * Assert strict equality
     * ```js
     * actual === expected
     * ```
     */
    equal(actual, expected, message) {
        if (!(actual === expected)) {
            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
            throw new AssertionError(`${message || "Expected values to be equal"}: ${actual} === ${expected}`);
        }
    },
    /**
     * Assert less than or equal
     * ```js
     * left <= right
     * ```
     */
    lte(left, right, message) {
        if (!(left <= right)) {
            throw new AssertionError(`${message || "Expected value to be lte"}: ${left} <= ${right}`);
        }
    },
    /**
     * Assert less than
     * ```js
     * left < right
     * ```
     */
    lt(left, right, message) {
        if (!(left < right)) {
            throw new AssertionError(`${message || "Expected value to be lt"}: ${left} < ${right}`);
        }
    },
    /**
     * Assert greater than or equal
     * ```js
     * left >= right
     * ```
     */
    gte(left, right, message) {
        if (!(left >= right)) {
            throw new AssertionError(`${message || "Expected value to be gte"}: ${left} >= ${right}`);
        }
    },
    /**
     * Assert greater than
     * ```js
     * left > right
     * ```
     */
    gt(left, right, message) {
        if (!(left > right)) {
            throw new AssertionError(`${message || "Expected value to be gt"}: ${left} > ${right}`);
        }
    },
};
export class AssertionError extends Error {
}
AssertionError.code = "ERR_ASSERTION";
